Release `LockData::dd_mutex` before calling `*_detected` functions

hebasto opened the pull request:
Both `double_lock_detected()` and `potential_deadlock_detected()` functions call `LogPrintf()` which in turn implies locking of the `Logger::m_cs` mutex. To avoid a deadlock, the latter must not have the `Mutex` type (see https://github.com/bitcoin/bitcoin/pull/16112).

With this PR the mentioned restriction has been lifted, and it is possible now to use our regular `Mutex` type for the `Logger::m_cs` mutex instead of a dedicated `StdMutex` type (not introducing that change here, as its diff is much bigger than a few lines, and the currently proposed diff seems valuable by itself).

**Note for reviewers:** Make sure the code is configured and built with `CPPFLAGS=-DDEBUG_LOCKORDER`.

vasild commented on src/sync.cpp:
```diff
@@ -184,6 +185,7 @@ static void push_lock(MutexType* c, const CLockLocation& locklocation)
             auto lock_stack_copy = lock_stack;
             lock_stack.pop_back();
+            lock.unlock();
             potential_deadlock_detected(p1, lockdata.lockorders[p2], lock_stack_copy);
```
Looking more carefully, `p1` is just a pair of `void*` that are never dereferenced, so no need to create a copy of that.

vasild reviewed (approved):
ACK 20647b8b805e96f18e6d122ada23fef363a96619

hebasto commented:
cc @ryanofsky

hebasto commented:
Rebased to refresh the CI.

vasild reviewed (approved):
ACK 38f93fe0cb680425f5fec7c794b39c0e1795f9dc

ryanofsky commented on src/sync.cpp:
```diff
@@ -172,6 +172,7 @@ static void push_lock(MutexType* c, const CLockLocation& locklocation)
             // same thread as that results in an undefined behavior.
             auto lock_stack_copy = lock_stack;
             lock_stack.pop_back();
+            lock.unlock();
```
In commit "Release `LockData::dd_mutex` before calling `*_detected` functions" (08cfe273ddf9bd7640456865926ec9d94b363520)

Would be good to have a comment saying why this is necessary. If I understand correctly it could say something like // Release dd_mutex before calling double_lock_detected to avoid deadlocks, so no other lock is ever acquired after dd_mutex, and lock order is consistent.

ryanofsky commented on src/sync.cpp:
```diff
@@ -182,9 +183,11 @@ static void push_lock(MutexType* c, const CLockLocation& locklocation)
         const LockPair p2 = std::make_pair(c, i.first);
         if (lockdata.lockorders.count(p2)) {
-            auto lock_stack_copy = lock_stack;
+            auto lock_stack_current = lock_stack;
```
In commit "Release `LockData::dd_mutex` before calling `*_detected` functions" (08cfe273ddf9bd7640456865926ec9d94b363520)

Previous name `lock_stack_copy` seems like it explains the purpose for this variable better than `lock_stack_current`. Also `lock_stack_copy` is the name used above and it would be nicer if double_lock_detected and potential_deadlock_detected cases were handled consistently I think, unless I am missing something.

ryanofsky reviewed (commented):
Code review ACK 38f93fe0cb680425f5fec7c794b39c0e1795f9dc. This seems like a good change, but I think it would be better if it had another commit getting rid of StdMutex/StdLockGuard. Otherwise nothing is really taking advantage of the change.

achow101 commented:
@hebasto Are you planning on addressing review feedback here?

***

> I think it would be better if it had another commit getting rid of StdMutex/StdLockGuard. Otherwise nothing is really taking advantage of the change.

I agree. The PR description sounded like this was going to be the ultimate change, but it does not appear to have actually been done.

vasild reviewed (approved):
ACK 38f93fe0cb680425f5fec7c794b39c0e1795f9dc

(repeat my earlier comment, DrahtBot was confused)

hebasto committed:
sync: Factor out `inconsistent_lock_order_detected` function

This simplifies enforcing the rule of not holding `LockData::dd_mutex`
across all `*_detected` functions in subsequent commits.

Additionally, the `[[noreturn]]` attribute has been added to the
`potential_deadlock_detected` and `double_lock_detected` functions.

hebasto committed:
sync: Release `LockData::dd_mutex` before calling `*_detected` functions

All `*_detected()` functions call `LogError()`, which requires acquiring
the `Logger::m_cs` mutex. To avoid a potential lock deadlock, the latter
must not have the `Mutex` type.

This change lifts the restriction on the `Logger::m_cs` type, allowing
it to be transitioned from `StdMutex` to our standard `Mutex` type in
subsequent commits.

hebasto committed:
logging: Migrate `Logger::m_cs` type from `StdMutex` to `Mutex`

There is no longer a need for the `StdMutex` type, which was
specifically implemented for use in the `Logger` class.

hebasto committed:
logging: Migrate `LogRateLimiter::m_mutex` type to `Mutex`

hebasto committed:
refactor: Remove unused `StdMutex` and `StdLockGuard`
